packages/opentelemetry-resources/src/platform/node/detectors/ProcessDetector.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Detector,
+  Resource,
+  PROCESS_RESOURCE,
+  ResourceDetectionConfigWithLogger,
+} from '../../../';
+import { ResourceAttributes } from '../../../types';
+
+/**
+ * ProcessDetector will be used to detect the resoureces related current process running
+ * and being instumented from the NodeJS Process module.
+ */
+class ProcessDetector implements Detector {
+  async detect(config: ResourceDetectionConfigWithLogger): Promise<Resource> {
+    const processResource: ResourceAttributes = {
+      [PROCESS_RESOURCE.PID]: process.pid || '',
+      [PROCESS_RESOURCE.NAME]: process.title || '',
+      [PROCESS_RESOURCE.COMMAND]: process.argv[1] || '',
+      [PROCESS_RESOURCE.COMMAND_LINE]: process.argv.join(' ') || '',
+    };
+    return this._getResourceAttributes(processResource, config);
+  }
+  /**
+   * Validates process resource attribute map from process varaibls
+   *
+   * @param processResource The unsantized resource attributes from process as key/value pairs.
+   * @param config: Config
+   * @returns The sanitized resource attributes.
+   */
+  private _getResourceAttributes(
+    processResource: ResourceAttributes,
+    config: ResourceDetectionConfigWithLogger
+  ) {
+    if (processResource[PROCESS_RESOURCE.PID] === '') {
+      config.logger.debug(
+        'ProcessDetector failed: unable to locate pid for currently running process'
+      );
+      return Resource.empty();
```

Comment:
Do we want to return an empty resource or just best-effort for the ones we can detect?

Reply:
There is condition, we return empty only when we don't meet specification conditions which is.

* PID is mandatory
* At least one of `process.executable.name`, `process.executable.path`, `process.command`, or `process.command_line` is required.

Other wise we return best effort

Reply:
I guess that's probably fine. I was thinking we should return best effort no matter what, but actually I'm pretty sure `process.pid` never fails. I just looked at the docs and didn't see any mention that it is possible to not get a number back.

Reply:
Agree it will never fail, but as per spec two members are mandatory so i suspect we should be adhering to that right ?

Reply:
There is no reason to have an impossible branch

Reply:
Makes sense, have removed it.